README.md

[![Dependency Status](https://david-dm.org/webplatform/mdn-compat-importer.png)](https://david-dm.org/webplatform/mdn-compat-importer)

# mdn-compat-importer

Reads MDN compatibility info and converts it from HTML to JSON.

***This tool isn't meant to be run as-is, it will exit with a warning by default!***

### Installation

  - Clone this repository (you should fork first, if you want to contribute)
  - Install [NodeJS](http://nodejs.org/), if you haven't already
  - Run `npm install -g grunt-cli` (you might need sudo/admin for this)
  - Run `npm install` from the directory to install the dependencies
  - Run `node index.js` to run the tool

### Current Issues

  - The internal JS object needs to be converted into something that represents the [Browser Compatibility Tables - Datamodel](http://www.ronaldmansveld.nl/webplatform/compat_tables_datamodel.html)
  - `Converter#extract()` needs more work to handle complex compat tables, like the one from [`<input>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#Browser_compatibility) and probably more compat tables in general
  - Modeled after the old script, pages are crawled according to tags (CSS, HTML, HTML5, API and WebAPI at the moment), but tag feeds are limited to 500 results and don't seem to support pagination (see the [docs](https://developer.mozilla.org/en-US/docs/Project:MDN/Tools/Feeds) and [code](https://github.com/mozilla/kuma/blob/master/apps/wiki/feeds.py)), might need a better way to discover pages, current stats are:
    - 1956 pages for those 5 tags (370 in HTML, 86 in HTML5)
    - 1576 pages left after removing duplicates
    - 846 pages have a compat section
  - Also, some pages have compat tables but don't represent a tag or property directly, e.g. [HTML/CORS_Enabled_Image](https://developer.mozilla.org/en-US/docs/HTML/CORS_Enabled_Image)

### Developing

A full cache is bundled, so no requests will be made to MDN. You can start working on the HTML to JS parser or the conversion to WPD format right away! If you have questions, post to the mailing list or catch me (frozenice) on IRC.

To test single pages, uncomment the stuff around line 50 and line 66 in `index.js`.

### Internal Object

Here are some examples of what the internal JS object looks like:

#### [HTML/Element/header](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/header#Browser_compatibility) (simple table)

```js
{ 'https://developer.mozilla.org/en-US/docs/Web/HTML/Element/header':
   { desktop:
      { 'Basic support':
         { Chrome: { normal: '5' },
           'Firefox (Gecko)': { normal: '4.0 (2.0)' },
           'Internet Explorer': { normal: '9.0' },
           Opera: { normal: '11.10' },
           Safari: { normal: '4.1' } } },
     mobile:
      { 'Basic support':
         { Android: { normal: '2.2' },
           'Firefox Mobile (Gecko)': { normal: '4.0 (2.0)' },
           'IE Mobile': { normal: '9.0' },
           'Opera Mobile': { normal: '11.0' },
           'Safari Mobile': { normal: '5.0 (iOS 4.2)' } } } } }
```

#### [HTML/Element/video](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video#Browser_compatibility) (table with several rows)

```js
{ 'https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video':
   { desktop:
      { 'Basic support':
         { Chrome: { normal: '3.0' },
           'Firefox (Gecko)': { normal: '3.5 (1.9.1)' },
           'Internet Explorer': { normal: '9.0' },
           Opera: { normal: '10.5' },
           Safari: { normal: '3.1' } },
        'autoplay attribute':
         { Chrome: { normal: '3.0' },
           'Firefox (Gecko)': { normal: '3.5 (1.9.1)' },
           'Internet Explorer': { normal: '9.0' },
           Opera: { normal: '10.5' },
           Safari: { normal: '3.1' } },
        'buffered attribute':
         { Chrome: { normal: '?' },
           'Firefox (Gecko)': { normal: '4.0 (2.0)' },
           'Internet Explorer': { normal: '?' },
           Opera: { normal: 'Yes' },
           Safari: { normal: '?' } },
        'controls attribute':
         { Chrome: { normal: '3.0' },
           'Firefox (Gecko)': { normal: '3.5 (1.9.1)' },
           'Internet Explorer': { normal: '9.0' },
           Opera: { normal: '10.5' },
           Safari: { normal: '3.1' } },
        'loop attribute':
         { Chrome: { normal: '3.0' },
           'Firefox (Gecko)': { normal: '11.0 (11.0)' },
           'Internet Explorer': { normal: '9.0' },
           Opera: { normal: '10.5' },
           Safari: { normal: '3.1' } },
        'muted attribute':
         { Chrome: { normal: '?' },
           'Firefox (Gecko)': { normal: '11.0 (11.0)' },
           'Internet Explorer': { normal: '?' },
           Opera: { normal: 'Yes' },
           Safari: { normal: '?' } },
        'played property':
         { Chrome: { normal: '?' },
           'Firefox (Gecko)': { normal: '15.0 (15.0)' },
           'Internet Explorer': { normal: '?' },
           Opera: { normal: 'Yes' },
           Safari: { normal: '?' } },
        'poster attribute':
         { Chrome: { normal: '3.0' },
           'Firefox (Gecko)': { normal: '3.6 (1.9.2)' },
           'Internet Explorer': { normal: '9.0' },
           Opera: { normal: '10.5' },
           Safari: { normal: '3.1' } },
        'preload attribute':
         { Chrome: { normal: '3.0' },
           'Firefox (Gecko)': { normal: '4.0 (2.0)' },
           'Internet Explorer': { normal: '9.0' },
           Opera: { normal: 'Yes' },
           Safari: { normal: '3.1' } },
        'src attribute':
         { Chrome: { normal: '3.0' },
           'Firefox (Gecko)': { normal: '3.5 (1.9.1)' },
           'Internet Explorer': { normal: '9.0' },
           Opera: { normal: '10.5' },
           Safari: { normal: '3.1' } },
        'crossorigin attribue':
         { Chrome: { normal: '?' },
           'Firefox (Gecko)': { normal: '12.0 (12.0)' },
           'Internet Explorer': { normal: '?' },
           Opera: { normal: '?' },
           Safari: { normal: '?' } } },
     mobile:
      { 'Basic support':
         { Android: { normal: '?' },
           'Firefox Mobile (Gecko)': { normal: '1.0 (1.0)' },
           'IE Mobile': { normal: '?' },
           'Opera Mobile': { normal: '?' },
           'Safari Mobile': { normal: '?' } },
        'autoplay attribute':
         { Android: { normal: '?' },
           'Firefox Mobile (Gecko)': { normal: '1.0 (1.0)' },
           'IE Mobile': { normal: '?' },
           'Opera Mobile': { normal: '?' },
           'Safari Mobile': { normal: 'iOS 6.0 only' } },
        'buffered attribute':
         { Android: { normal: '?' },
           'Firefox Mobile (Gecko)': { normal: '4.0 (2.0)' },
           'IE Mobile': { normal: '?' },
           'Opera Mobile': { normal: '?' },
           'Safari Mobile': { normal: '?' } },
        'controls attribute':
         { Android: { normal: '?' },
           'Firefox Mobile (Gecko)': { normal: '1.0 (1.0)' },
           'IE Mobile': { normal: '?' },
           'Opera Mobile': { normal: '?' },
           'Safari Mobile': { normal: '?' } },
        'loop attribute':
         { Android: { normal: '?' },
           'Firefox Mobile (Gecko)': { normal: '11.0 (11.0)' },
           'IE Mobile': { normal: '?' },
           'Opera Mobile': { normal: '?' },
           'Safari Mobile': { normal: '?' } },
        'muted attribute':
         { Android: { normal: '?' },
           'Firefox Mobile (Gecko)': { normal: '11.0 (11.0)' },
           'IE Mobile': { normal: '?' },
           'Opera Mobile': { normal: '?' },
           'Safari Mobile': { normal: '?' } },
        'played property':
         { Android: { normal: '?' },
           'Firefox Mobile (Gecko)': { normal: '15.0 (15.0)' },
           'IE Mobile': { normal: '?' },
           'Opera Mobile': { normal: '?' },
           'Safari Mobile': { normal: '?' } },
        'poster attribute':
         { Android: { normal: '?' },
           'Firefox Mobile (Gecko)': { normal: '1.0 (1.0)' },
           'IE Mobile': { normal: '?' },
           'Opera Mobile': { normal: '?' },
           'Safari Mobile': { normal: '?' } },
        'preload attribute':
         { Android: { normal: '?' },
           'Firefox Mobile (Gecko)': { normal: '4.0 (2.0)' },
           'IE Mobile': { normal: '?' },
           'Opera Mobile': { normal: '?' },
           'Safari Mobile': { normal: '?' } },
        'src attribute':
         { Android: { normal: '?' },
           'Firefox Mobile (Gecko)': { normal: '1.0 (1.0)' },
           'IE Mobile': { normal: '?' },
           'Opera Mobile': { normal: '?' },
           'Safari Mobile': { normal: '?' } },
        'crossorigin attribute':
         { Android: { normal: '?' },
           'Firefox Mobile (Gecko)': { normal: '12.0 (12.0)' },
           'IE Mobile': { normal: '?' },
           'Opera Mobile': { normal: '?' },
           'Safari Mobile': { normal: '?' } } } } }
```

#### [CSS/@keyframes](https://developer.mozilla.org/en-US/docs/Web/CSS/@keyframes#Browser_Compatibility) (table with prefixes)

```js
{ 'https://developer.mozilla.org/en-US/docs/Web/CSS/@keyframes':
   { desktop:
      { 'Basic support':
         { Chrome: { prefix: '(Yes)' },
           'Firefox (Gecko)': { prefix: '5.0 (5.0)', normal: '16.0 (16.0)' },
           'Internet Explorer': { normal: '10' },
           Opera: { prefix: '12', normal: '12.10' },
           'Safari (WebKit)': { prefix: '4.0' } } },
     mobile:
      { 'Basic support':
         { Android: { prefix: '(Yes)' },
           'Firefox Mobile (Gecko)': { prefix: '5.0 (5.0)', normal: '16.0 (16.0)' },
           'IE Phone': { normal: '?' },
           'Opera Mobile': { normal: '?' },
           'Safari Mobile': { normal: '?' } } } } }
```

---

This project was generated using [generator-node-adv](https://github.com/webplatform/generator-node-adv), check there for general NodeJS docs and information about the project's structure.


CHANGELOG.md
# Changelog

#### 0.0.1
  - Does what the old script did, with some modifications

#### 0.0.0
  - Initial project setup


test/init.js
var path = require('path');
var chai = require('chai');
var istanbul = require('istanbul'),
  hook = istanbul.hook,
  Instrumenter = istanbul.Instrumenter,
  instrumenter = new Instrumenter();
var cwd = process.cwd();
var cwdNodeModules = path.resolve(cwd, 'node_modules');
var cwdTest = path.resolve(cwd, 'test');
var matcher = function(file) {
  if (file.indexOf(cwd) !== 0 || file.indexOf(cwdNodeModules) === 0 || file.indexOf(cwdTest) === 0) {
    return false;
  }
  return true;
};
var transformer = instrumenter.instrumentSync.bind(instrumenter);
hook.hookRequire(matcher, transformer);

global.assert = chai.assert;
global.expect = chai.expect;
chai.should();


test/Reader/index.js
describe('The Reader module', function() {
  it('should be exposed', function() {
    require('../../index.js').Reader.should.exist;
  });
});


test/Converter/index.js
describe('The Converter module', function() {
  it('should be exposed', function() {
    require('../../index.js').Converter.should.exist;
  });
});


index.js
// mdn-compat-importer
var fs = require('fs');
var mkdirp = require('mkdirp');
var Q = require('q');
var Reader = require('./lib/Reader.js');
var Converter = require('./lib/Converter.js');
var _ = require('lodash-node');

if (fs.existsSync('.disabled')) {
  console.log('WARNING');
  console.log('Make sure you understand the implications of this tool before running!');
  console.log('Remove the .disabled file when you do.');
  return;
}

mkdirp.sync('data/cache');

var reader = new Reader();
var converter = new Converter();

var tags = ['CSS', 'HTML', 'HTML5', 'API', 'WebAPI'];

Q.
  // process each tag and wait for everything to finish
  all(tags.map(function(tag) {
    // get a promise for this tag
    return reader.processFeed(tag);
  })).

  // find out how the compat section is named for each page
  then(function() {
    console.log('retrieving compat section ids for ' + _.size(reader.links) + ' pages');
    return reader.getSectionNames();
  }).

  // fetch the compat sections
  then(function(links) {
    console.log('retrieving compat sections for ' + _.size(links) + ' pages');
    return reader.fetchSections();
  }).

  // we are only interested in pages with at least something in their compat tables section
  then(function(pages) {
    var pagesWithCompatTables = _.pick(pages, function(html) {
      return html && html.length > 10;
    });
    console.log('found ' + _.size(pagesWithCompatTables) + ' pages with compat tables');
    return pagesWithCompatTables;

    // use these to test specific pages, comment the return directly above this line
    //return _.pick(pagesWithCompatTables, function(value, key) { return key === 'https://developer.mozilla.org/en-US/docs/Web/HTML/Element/header'; }); // simple table
    //return _.pick(pagesWithCompatTables, function(value, key) { return key === 'https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video'; }); // table with several rows
    //return _.pick(pagesWithCompatTables, function(value, key) { return key === 'https://developer.mozilla.org/en-US/docs/Web/CSS/@keyframes'; }); // table with prefixes
    //return _.pick(pagesWithCompatTables, function(value, key) { return key === 'https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input'; }); // very complex table
    //
    // if you want to grab several pages, use this:
    //return _.pick(pagesWithCompatTables,
    //  'https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video',
    //  'https://developer.mozilla.org/en-US/docs/Web/CSS/@keyframes'
    //);
  }).

  // convert the compat tables to JSON
  then(function(pages) {
    return _.mapValues(pages, function(page) {
      return converter.extract(page);
    });
  }).

  // transform to WPD format
  then(function(tables) {
    //console.log(require('util').inspect(tables, { depth: 4 })); // uncomment to see the object, single page mode from above recommended
    var wpd = {};
    _.each(tables, function(table, url) {
      var compat = converter.convert(table, url);
      wpd[compat.uuid] = compat.data;
    });
    return wpd;
  }).

  // save to disk
  then(function(wpd) {
    fs.writeFileSync('data/compat-mdn.json', JSON.stringify(wpd));
    console.log('data saved');
  }).

  // this will throw if something went wrong
  done();


Gruntfile.js
var fs = require('fs');
var cp = require('child_process');
var rimraf = require('rimraf');
var istanbul = require('istanbul');

var LINT = true;
var COVERALLS = false;

module.exports = function(grunt) {
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    jshint: {
      options: { // see http://www.jshint.com/docs/options/
        node: true,
        esnext: true,
        bitwise: true,
        camelcase: true,
        curly: true,
        eqeqeq: true,
        immed: true,
        indent: 2,
        latedef: true,
        newcap: true,
        noarg: true,
        undef: true,
        trailing: true,
        smarttabs: true
      },
      build: ['Gruntfile.js'],
      lib: ['index.js', 'lib/**/*.js'],
      test: {
        options: {
          '-W030': true, // Expected an assignment or function call and instead saw an expression: ...to.be.true
          globals: {
            describe: true,
            it: true,
            assert: true,
            expect: true,
            before: true,
            after: true,
            beforeEach: true,
            afterEach: true
          }
        },
        files: {
          src: ['test/**/*.js']
        }
      }
    },
    mochaTest: {
      lib: {
        options: {
          ui: 'bdd',
          require: ['test/init.js']
        },
        src: ['test/**/*.js', '!test/init.js', '!test/_fixtures/**']
      }
    }
  });

  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-mocha-test');

  grunt.registerTask('test', 'Build, generate coverage, run tests.', function() {
    if (fs.existsSync('test-results')) {
      rimraf.sync('test-results');
    }
    fs.mkdirSync('test-results');

    process.env.multi = 'spec=- mocha-slow-reporter=test-results/slow.txt';
    grunt.config.set('mochaTest.lib.options.reporter', 'mocha-multi');

    // needed so that mocha-multi doesn't kill the process
    var program = require('mocha/node_modules/commander');
    program.name = 'mocha';
    program.exit = false;

    grunt.task.run(['build', 'mochaTest', 'covreport']);
    if (process.env.TRAVIS) {
      grunt.config.set('mochaTest.lib.options.reporter', 'mocha-unfunk-reporter');
      if (COVERALLS) {
        grunt.task.run(['coveralls']);
      }
    }
  });

  grunt.registerTask('covreport', 'Generate coverage reports.', function() {
    var collector = new istanbul.Collector();
    collector.add(global.__coverage__);
    var reports = [
      istanbul.Report.create('lcovonly', { dir: 'test-results' })
    ];
    if (process.env.TRAVIS) {
      // show per-file coverage stats in the Travis console
      reports.push(istanbul.Report.create('text'));
    } else {
      // produce HTML when not running under Travis
      reports.push(istanbul.Report.create('html', { dir: 'test-results' }));
    }
    reports.push(istanbul.Report.create('text-summary'));
    reports.forEach(function(report) {
      report.writeReport(collector, true);
    });
  });

  grunt.registerTask('coveralls', 'Push to Coveralls.', function() {
    this.requires('mochaTest');
    var done = this.async();
    cp.exec('cat ./test-results/lcov.info | ./node_modules/.bin/coveralls', { cwd: './' }, function(err, stdout, stderr) {
      grunt.log.writeln(stdout + stderr);
      done(err);
    });
  });

  grunt.registerTask('build', function() {
    if (LINT) {
      grunt.task.run('jshint');
    }
  });

  grunt.registerTask('default', ['build']);
};


lib/Reader.js
var fs = require('fs');
var request = require('superagent');
var Q = require('q');
var _ = require('lodash-node');

function Reader() {
  this.links = {}; // url -> 'Browser_compatibility'
  this.pages = {}; // url -> compat section as HTML

  // try to load a cache file
  this.cachePath = 'data/cache/pages.json';
  this.cache = {};
  if (fs.existsSync(this.cachePath)) {
    this.cache = JSON.parse(fs.readFileSync(this.cachePath, {
      encoding: 'utf8'
    }));
    console.log('using cache file containing ' + _.size(this.cache) + ' pages');
  }
}

/**
 * fetches a feed either from cache or from the web
 * @param  {String}   tag the tag to fetch
 * @param  {Function} cb  a callback to notify when finished
 */
Reader.prototype.getFeed = function(tag, cb) {
  // use the cached version, if present
  var cachePath = 'data/cache/tag-' + tag.toLowerCase() + '.json';
  if (fs.existsSync(cachePath)) {
    var content = fs.readFileSync(cachePath, {
      encoding: 'utf8'
    });
    var obj = JSON.parse(content);
    console.log('using cached feed containing ' + obj.length + ' links for tag: ' + tag);
    cb(null, obj);
    return;
  }

  // else fetch a fresh copy
  var url = 'https://developer.mozilla.org/en-US/docs/feeds/json/tag/' + tag;
  console.log('fetching live feed for tag: ' + tag);
  request.get(url).end(function(err, res) {
    // check for errors
    if (err) {
      console.log('error requesting feed for tag: ' + tag);
      cb(err);
      return;
    } else if (res.error) {
      console.log('unexpected feed response for tag: ' + tag);
      cb(res.error);
      return;
    }

    // everything ok!
    console.log(res.body.length + ' pages for tag: ' + tag);
    fs.writeFileSync(cachePath, JSON.stringify(res.body));
    cb(null, res.body);
  });
};

/**
 * extracts the page links from a feed
 * @param  {Object} feed a feed
 */
Reader.prototype.parseFeed = function(feed) {
  // feed will be an array of:
  // {
  //   pubdate: "2014-01-11T00:45:36-08:00"
  //   author_link: "https://developer.mozilla.org/en-US/profiles/teoli"
  //   title: "text-decoration-style"
  //   author_name: "teoli"
  //   author_avatar: "https://secure.gravatar.com/avatar/5217c19185715da9fc4e43afa63eeb8a?s=220&r=pg&d=https%3A%2F%2Fdeveloper.cdn.mozilla.net%2Fmedia%2F%2Fimg%2Favatar.png"
  //   link: "https://developer.mozilla.org/en-US/docs/Web/CSS/text-decoration-style"
  //   categories: {
  //     CSS3: "https://developer.mozilla.org/en-US/docs/tag/CSS3"
  //     Web: "https://developer.mozilla.org/en-US/docs/tag/Web"
  //     CSS Property: "https://developer.mozilla.org/en-US/docs/tag/CSS%20Property"
  //     Reference: "https://developer.mozilla.org/en-US/docs/tag/Reference"
  //     Référence: "https://developer.mozilla.org/en-US/docs/tag/R%C3%A9f%C3%A9rence"
  //     NeedsLiveSample: "https://developer.mozilla.org/en-US/docs/tag/NeedsLiveSample"
  //     css3-text: "https://developer.mozilla.org/en-US/docs/tag/css3-text"
  //     Layout: "https://developer.mozilla.org/en-US/docs/tag/Layout"
  //     CSS: "https://developer.mozilla.org/en-US/docs/tag/CSS"
  //   }
  // }
  feed.forEach(function(item) {
    // avoid duplicates when fetching multiple feeds
    this.links[item.link] = null;
  }.bind(this));
};

/**
 * fetch and parse a feed
 * @param  {String}   tag the tag to fetch and process
 * @return {Promise}      a promise which will be resolved when finished
 */
Reader.prototype.processFeed = function(tag, cb) {
  var defer = Q.defer();
  this.getFeed(tag, function(err, feed) {
    if (err) {
      defer.reject(err);
    } else {
      this.parseFeed(feed);
      defer.resolve(this.links);
    }
  }.bind(this));
  return defer.promise;
};

/**
 * reads a page's metadata and figures out how exactly the compat section is named
 * this step is needed, because sometimes it's 'Browser_Compatibility' and sometimes it's 'Browser_compatibility'
 */
Reader.prototype.getSectionNames = function() {
  var sleepInterval = 1000; // how many ms to wait between requests
  var defer = Q.defer();
  var todo = _.clone(this.links);

  // create the worker
  var processSingleLink = function() {
    var size = _.size(todo);
    if (size > 0) {
      // there is more work
      if (size % 50 === 0) {
        console.log(size + ' section ids left...');
      }

      // grab the first item from the queue
      var url = _.keys(todo)[0];
      delete todo[url];

      // check the cache
      if (_(this.cache).has(url)) {
        // we can start the next work item immediately, because no request occured
        setImmediate(processSingleLink);
      } else {
        // fetch it from the web
        request.get(url + '$json').end(function(res) {
          var compatSection = _.find(res.body.sections, function(section) {
            return section.id.toLowerCase() === 'browser_compatibility';
          });
          if (compatSection) {
            this.links[url] = compatSection.id;
          }

          // schedule the next work item
          setTimeout(processSingleLink, sleepInterval);
        }.bind(this));
      }
    } else {
      // finish when everything is done
      defer.resolve(_.pluck(this.links));
    }
  }.bind(this);

  // start the worker
  setImmediate(processSingleLink);

  return defer.promise;
};

/**
 * goes through all links and fetches the compat section for each
 * @return {Promise} a promise which will be resolved when finished
 */
Reader.prototype.fetchSections = function() {
  var sleepInterval = 1000; // how many ms to wait between requests
  var defer = Q.defer();
  var todo = _.clone(this.links);

  // create the worker
  var processSinglePage = function() {
    var size = _.size(todo);
    if (size > 0) {
      // there is more work
      if (size % 50 === 0) {
        console.log(size + ' pages left...');
      }

      // grab the first item from the queue
      var url = _.keys(todo)[0];
      var section = todo[url];
      delete todo[url];

      // check the cache
      if (_(this.cache).has(url)) {
        this.pages[url] = this.cache[url];
        // we can start the next work item immediately, because no request occured
        setImmediate(processSinglePage);
      } else if (!section) {
        // there is no compat section, mark it as null for the cache
        this.pages[url] = null;
        // ignore pages without compat sections
        setImmediate(processSinglePage);
      } else {
        // fetch it from the web
        request.get(url + '?raw&macros&section=' + section).end(function(res) {
          // this will also cache pages without a compat section, so they won't be requested again
          this.pages[url] = res.text;

          // schedule the next work item
          setTimeout(processSinglePage, sleepInterval);
        }.bind(this));
      }
    } else {
      // finish when everything is done
      _.assign(this.cache, this.pages); // put new pages into the cache
      fs.writeFileSync(this.cachePath, JSON.stringify(this.cache));
      defer.resolve(this.pages);
    }
  }.bind(this);

  // start the worker
  setImmediate(processSinglePage);

  return defer.promise;
};

module.exports = Reader;


lib/Converter.js
var $ = require('cheerio');
var _ = require('lodash-node');

function Converter() {
}

/**
 * extracts an JS object from a HTML representation, it will be in the following format
 * version values can also be '?', 'Yes', '(Yes)' 'iOS 6.0 only', '5.0 (iOS 4.2)', '---', 'Not supported' etc.
 * desktop and mobile can be missing
 * normal and prefix can be missing
 * {
 *   desktop: {
 *     'Basic support': {
 *       Chrome: {
 *         normal: '3.0',
 *         prefix: '(Yes)'
 *       },
 *       'Firefox (Gecko)': {
 *         normal: '16.0 (16.0)',
 *         prefix: '5.0 (5.0)'
 *       },
 *       // ...
 *     },
 *     'autoplay attribute': {
 *       // see above
 *     },
 *     // ...
 *   },
 *   mobile: {
 *     // see above
 *   }
 * }
 * @param  {String} html the HTML to parse
 * @return {Object}      the compatibility info
 */
Converter.prototype.extract = function(html) {
  // see here for help with cheerio:
  // https://github.com/MatthewMueller/cheerio
  var table = {};

  function parseTable($ele) {
    var table = {};
    var browsers = $('th', $ele).toArray().map(function(th) {
      return $(th).text();
    }).slice(1);
    $('tr:not(:first-child)', $ele).toArray().forEach(function(tr) {
      var feature = $('td:first-child', $(tr)).text();
      var versions = $('td:not(:first-child)', $(tr)).toArray().map(function(td) {
        var version = {};

        var html = $(td).html();
        // for prefixes td will be something like this:
        // 5.0 (5.0)<span class="inlineIndicator prefixBox prefixBoxInline" title="prefix">-moz</span><br>16.0 (16.0)
        // or:
        // <span style="color: #888;" title="Please...">(Yes)</span><span class="inlineIndicator prefixBox prefixBoxInline" title="prefix">-webkit</span>
        // so we always split by <br>
        var parts = html.split('<br>');
        parts.forEach(function(part) {
          var $part = $.load('<div>' + part + '</div>'); // wrap into a <div>, because part can be text-only

          // clean the text
          var $copy = $.load($part.html());
          $copy('a').remove();
          $copy('.prefixBox').remove();
          $copy('.unimplemented').remove();
          var cleanedText = $copy(':root').text().trim();

          // check if there is a span with class prefixBox in the part
          if ($part('.prefixBox').length > 0) {
            // looks like a prefixed version!
            version.prefix = cleanedText;
          } else {
            // probably normal version
            version.normal = cleanedText;
          }
        });

        return version;
      });
      table[feature] = _.zipObject(browsers, versions);
    });
    return table;
  }

  var desktop = $('#compat-desktop', html);
  if (desktop) {
    table.desktop = parseTable(desktop);
  }

  var mobile = $('#compat-mobile', html);
  if (mobile) {
    table.mobile = parseTable(mobile);
  }

  return table;
};

/**
 * converts an object from {@link Converter#extract} to WPD's format
 * see http://www.ronaldmansveld.nl/webplatform/compat_tables_datamodel.html
 * @param  {Object} table object from {@link Converter#extract}
 * @param  {String} url   where this data comes from
 * @return {Object}       compatibility info in WPD format
 */
Converter.prototype.convert = function(table, url) {
  var compat = {};

  // TODO, replace with proper conversion!
  compat.uuid = url;
  compat.data = table;

  return compat;
};

module.exports = Converter;
